### PR TITLE
*: Support reviewing audiences of token

### LINKED ISSUE #18: Support audience validation
Kubernetes now has a TokenRequest API, through which a ServiceAccount can create new tokens, that are scoped to audiences. For example this can be used by Prometheus to create a new token, that is purely meant for talking to kubelets. This is desirable, because otherwise the kubelet could simply impersonate the Prometheus server by stealing the ServiceAccount token it handed to it in plaintext.

Passing audiences to the TokenReview API is not implemented yet, so implementing support for this in the kube-rbac-proxy is currently blocked on this merging in Kubernetes. (RE: https://github.com/kubernetes/kubernetes/pull/62692)

### PR DESCRIPTION
**Pull request by brancz**
This adds the ability to verify the audience of a token, so clients can use audience scoped tokens, as opposed to essentially allowing the kube-rbac-proxy to impersonate the requesting client using its serviceaccount token.

Closes #18 

@s-urbaniak @paulfantom @LiliC @metalmatze

### CODE REVIEW

**Comment by s-urbaniak** [comment]
do you mind to put the vendoring into a separate commit?

**Comment by brancz** [comment]
Done.

File: examples/non-resource-url-token-request/non-resource-url/README.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+# non-resource-url example
```
Comment (s-urbaniak) [inline examples/non-resource-url-token-request/non-resource-url/README.md]: hmm, how does this differ from `examples/non-resource-url-token-request/README.md`? Or is this duplication just accidentally?

File: examples/non-resource-url/non-resource-url/README.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
+# non-resource-url example
```
Comment (s-urbaniak) [inline examples/non-resource-url/non-resource-url/README.md]: same question as above: how do these example differ?

**Comment by s-urbaniak** [comment]
@brancz looking good so far, just a couple of questions:
- the change commit contains vendoring code, do you mind to keep it separate?
- I _think_ there are duplicate readmes, or not?
- How does audiences relate to non-resource URL protection?

File: examples/non-resource-url-token-request/non-resource-url/README.md (hunk 1 above)
Comment (brancz) [reply examples/non-resource-url-token-request/non-resource-url/README.md]: whops I haven't finished this yet it seems

**Comment by brancz** [comment]
> the change commit contains vendoring code, do you mind to keep it separate?

Done, sorry about that.

> I think there are duplicate readmes, or not?

Fixed this.

> How does audiences relate to non-resource URL protection?

It doesn't really, I just wanted to provide some example showing the functionality. I added a note at the top of the example readme to make sure that's clearer:

> Token audience reviews can be combined with any of the other features of kube-rbac-proxy, such as reviewing permission for specific resources, this example merely chooses to show the functionality using a non-resource-url permission.

**Comment by metalmatze** [comment]
Are you still waiting for reviews or something else?

**Comment by brancz** [comment]
Waiting for reviews :)

File: README.md [hunk 3]
```diff
@@ -28,18 +28,22 @@ All command line flags:
 ```txt
 $ kube-rbac-proxy -h
 Usage of _output/linux/amd64/kube-rbac-proxy:
+      --add_dir_header                              If true, adds the file directory to the header
```
Comment (s-urbaniak) [inline README.md]: I am wondering if this is used at all?

File: README.md [hunk 4]
```diff
@@ -48,15 +52,18 @@ Usage of _output/linux/amd64/kube-rbac-proxy:
...
+      --skip_headers                                If true, avoid header prefixes in the log messages
+      --skip_log_headers                            If true, avoid headers when opening log files
       --stderrthreshold severity                    logs at or above this threshold go to stderr (default 2)
       --tls-cert-file string                        File containing the default x509 Certificate for HTTPS. (CA cert, if any, concatenated after server cert)
       --tls-cipher-suites strings                   Comma-separated list of cipher suites for the server. Values are from tls package constants (https://golang.org/pkg/crypto/tls/#pkg-constants). If omitted, the default Go cipher suites will be used
       --tls-min-version string                      Minimum TLS version supported. Value must match version names from https://golang.org/pkg/crypto/tls/#pkg-constants. (default "VersionTLS12")
       --tls-private-key-file string                 File containing the default x509 private key matching --tls-cert-file.
+      --tls-reload-interval duration                The interval at which to watch for TLS certificate changes, by default set to 1 minute. (default 1m0s)
```
Comment (s-urbaniak) [inline README.md]: good catch :+1:

File: examples/non-resource-url-token-request/README.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
+# non-resource-url-token-request example
+
+> Token audience reviews can be combined with any of the other features of kube-rbac-proxy, such as reviewing permission for specific resources, this example merely chooses to show the functionality using a non-resource-url permission.
+
+This examples is in essence similar to the [non-resource-url](../non-resource-url/) example, with the key difference, that this example requires the ServiceAccount token sent by a client must be scoped to the `kube-rbac-proxy.default.svc` audience. In this example the scoped ServiceAccount token is obtained via a projected volume and mounted into the client container from where it can be consumed. The reasoning here is that scroped tokens cannot be used to impersonate an entity by re-using the token to perform a request against the Kubernetes API itself.
```
Comment (s-urbaniak) [inline examples/non-resource-url-token-request/README.md]: s/scroped/scoped

File: README.md (hunk 3 above)
Comment (brancz) [reply README.md]: It’s from klog

File: README.md (hunk 4 above)
Comment (brancz) [reply README.md]: This actually came in as I regenerated the help text

File: examples/non-resource-url-token-request/README.md (hunk 5 above)
Comment (brancz) [reply examples/non-resource-url-token-request/README.md]: Fixed.

**Comment by brancz** [comment]
I split the PR into multiple commits, which should make it very easy to review now, as the actual code changes (in the second commit) are minimal and are mainly upgrading to using the v1 APIs. The rest are tests and docs (95% of this PR).

File: pkg/authn/delegating.go [hunk 6]
```diff
@@ -22,21 +22,33 @@ import (
...
 	}
 
+	var (
+		p   authenticatorfactory.CAContentProvider
+		err error
+	)
+	if len(authn.X509.ClientCAFile) > 0 {
+		p, err = dynamiccertificates.NewStaticCAContentFromFile(authn.X509.ClientCAFile)
+		if err != nil {
+			return nil, err
```
Comment (s-urbaniak) [inline pkg/authn/delegating.go]: just a nit: let's wrap this err, so context is visible in log lines.

**Comment by s-urbaniak** [comment]
this lgtm :tada: I really just have one nit regarding error wrapping, else this looks great. A small code footprint, but great security improvement :+1:

**Pull request merged**